### PR TITLE
oidcccl: Add configurable timeout to HTTP client

### PR DESCRIPTION
**Pull request by pritesh-lahoti**
Currently, the HTTP client used by the code in the OIDC feature does
not respect any timeouts we might have set on our default HTTP client.
The `go-oidc` and `oauth2` packages require a custom HTTP client to be
set within the context. This change introduces a cluster setting for
the timeout duration and sets its value in the HTTP client.

Epic: CRDB-38416

Release note (enterprise change, ops change): A new public cluster
setting `server.oidc_authentication.client.timeout` has been
introduced to capture the HTTP client timeout for external calls made
during OIDC authentication.

### CODE REVIEW

**Review by souravcrl — COMMENTED** [review]
nit: currently we provide full package name for commit/PR header, i.e. oidcccl: HTTP client should respect timeouts

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125767#-:-O-kIzoX-KqJp07REUA4:b-o5mob4)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian)

**Review by pritesh-lahoti — COMMENTED** [review]
Addressed.
Have also updated the same on the GitHub issue: https://github.com/cockroachdb/cockroach/issues/123605

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125767#-:-O-khFc08QHOgpa3B54i:bss855h)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian)

**Review by dhartunian — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125767)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pritesh-lahoti and @souravcrl)

___
*[`pkg/ccl/oidcccl/authentication_oidc.go` line 262 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/125767#-O-plw4-4N_1neeQPfR5:-O-plw4-4N_1neeQPfR6:b-owr62q) ([raw file](https://github.com/cockroachdb/cockroach/blob/67dccf7ead44a58ec7db12b5e14ef74cb3323283/pkg/ccl/oidcccl/authentication_oidc.go#L262)):*
> ```Go
> 
> 	// Set the HTTP client in the context, as required by the `go-oidc` and `oauth2` packages.
> 	octx := oidc.ClientContext(ctx, conf.httpClient.Client)
> ```

The manager is created in this function and then used to make the `Exchange` and `Verify` calls. What do you think of embedding the httpClient into the `oidcManager` struct so that it can be injected into the context in the `oidcManager` method implementations instead of outside. Then the caller of the `oidcManager` doesn't have to worry about this at all.
___
*[`pkg/ccl/oidcccl/authentication_oidc_test.go` line 73 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/125767#-O-ptMLD2guGQj2lDP-R:-O-po4cX8J5JQXg1hykd:b-qrsvcp) ([raw file](https://github.com/cockroachdb/cockroach/blob/67dccf7ead44a58ec7db12b5e14ef74cb3323283/pkg/ccl/oidcccl/authentication_oidc_test.go#L73)):*
> ```Go
> 
> func (m mockOidcManager) Verify(ctx context.Context, s string) (*oidc.IDToken, error) {
> 	httpClient, exists := ctx.Value(oauth2.HTTPClient).(*http.Client)
> ```

Consider testing for the timeout directly with a real http server.

You can use `s := httptest.NewServer(...)` to spin up a server on a random port and then call `s.URL` to get the path to pass to the `NewOIDCManager` constructor. It should be possible for the constructor to return within a timeout, even if the server handler hangs forever.

**Review by pritesh-lahoti — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125767)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @souravcrl)

___
*[`pkg/ccl/oidcccl/authentication_oidc.go` line 262 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/125767#-O-plw4-4N_1neeQPfR5:-O-uPtxP0Ozh6bRxJ0oW:baqjo34) ([raw file](https://github.com/cockroachdb/cockroach/blob/67dccf7ead44a58ec7db12b5e14ef74cb3323283/pkg/ccl/oidcccl/authentication_oidc.go#L262)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

The manager is created in this function and then used to make the `Exchange` and `Verify` calls. What do you think of embedding the httpClient into the `oidcManager` struct so that it can be injected into the context in the `oidcManager` method implementations instead of outside. Then the caller of the `oidcManager` doesn't have to worry about this at all.
</blockquote></details>

Sounds good to me. Have addressed.
___
*[`pkg/ccl/oidcccl/authentication_oidc_test.go` line 73 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/125767#-O-ptMLD2guGQj2lDP-R:-O-vKez-4a_JOJCQBdbj:b-ynoyir) ([raw file](https://github.com/cockroachdb/cockroach/blob/67dccf7ead44a58ec7db12b5e14ef74cb3323283/pkg/ccl/oidcccl/authentication_oidc_test.go#L73)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Consider testing for the timeout directly with a real http server.

You can use `s := httptest.NewServer(...)` to spin up a server on a random port and then call `s.URL` to get the path to pass to the `NewOIDCManager` constructor. It should be possible for the constructor to return within a timeout, even if the server handler hangs forever.
</blockquote></details>

Have added a test as suggested.

**Review by dhartunian — CHANGES_REQUESTED** [review]
I realize that I wrote the original ticket but I think my advice was misguided. Using a 3 second timeout here seems a bit brittle. The customer has no way to change this and I could imagine OIDC operations being a bit slow if they involve a bunch of intranet network hops at the customer.

Today, there is no timeout set on the default HTTP client so our code is actually quite open to a slow network on the customer side. I could see support tickets coming through with intermittent OIDC login problems from customers who suddenly start to experience timeouts after upgrading to a version with contained this PR.

Can you make the timeout a configurable cluster setting like the other OIDC settings? Then it can have a default of, perhaps, 30 seconds and be configurable by the customer if necessary. The desired outcome is that our HTTP request fails _eventually_ instead of waiting forever.

> Reviewed 2 of 4 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125767#-:-O0AxxdvDF4_evGc5GP1:bfvb1z4)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @souravcrl)

**Review by pritesh-lahoti — COMMENTED** [review]
Sounds fine to me.
I was wondering to reuse this cluster setting across the OIDC and JWT flows, and was looking at an appropriate place for setting this up. The oidcccl package already imports into jwtauthccl, so I was hoping to put this cluster setting within jwtauthccl.
As far as the name of the cluster setting is concerned, I am proposing: `server.authentication.external_calls.timeout`.

Feel free to suggest on the above proposal, and do let me know if we need to involve anyone else in making this decision. Thank you!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125767#-:-O0NAycR15adRYry9DIY:b-xnuejm)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @souravcrl)

**Review by dhartunian — COMMENTED** [review]
I'd use `server.authentication.client.timeout`, but your option sounds fine as well if you prefer it. To me, the namespacing is most important. I wish we could avoid adding an `.authentication` namespace to the settings, but agree with the motivation here.

One thing to consider is if the name will survive as we add more auth methods. I think it will. Maybe the next auth method can have settings under `server.authentication.<method>.<config var>` instead of how we did oidc/jwt so that things can line up a bit beter.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125767#-:-O0UmdP62rHbsYyW2rZl:b-1klhgm)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @souravcrl)

**Review by pritesh-lahoti — COMMENTED** [review]
> The oidcccl package already imports into jwtauthccl, so I was hoping to put this cluster setting within jwtauthccl.

This seemed a bit odd, as every other package would need to import into jwtauthccl, which is undesirable.
The other option was to place the cluster setting under `pkg/security` but seemed like a retrofit.

After discussing with @souravcrl, we decided to have specific timeout related settings for JWT and OIDC methods (and other auth methods in the future - e.g. LDAP). This should give more control to the customers, as each auth method may operate on a different server and protocol.

To summarize, we are leaning on to use `server.oidc_authentication.client_timeout` and `server.jwt_authentication.client_timeout`.
@dhartunian  Let us know if that sounds good to you, or in case you have any suggestions.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125767#-:-O0h3FS10MQGGkGPq8HX:b-tmfehg)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @souravcrl)

**Review by pritesh-lahoti — COMMENTED** [review]
Had intended to tag @dhartunian above.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125767#-:-O0hOkIhEVe1BfxLqhHC:b7ap34z)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @souravcrl)

**Review by pritesh-lahoti — COMMENTED** [review]
Have introduced a new cluster setting `server.oidc_authentication.client_timeout`, as discussed above.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125767#-:-O0izNCtE5NXDiCcfwK9:b-f5sa7j)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @souravcrl)

**Review by dhartunian — COMMENTED** [review]
> Reviewed 1 of 4 files at r4, 5 of 6 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125767)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pritesh-lahoti and @souravcrl)

___
*[`-- commits` line 2 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/125767#-O0nkQJG4vgSICLCxUfd):*
nit: just a stylistic nit that the title is a little vague. I'd use something like: `add configurable timeouts to HTTP clients`.
___
*[`-- commits` line 12 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/125767#-O0nkI0i9rNPc_1AyjmH):*
nit: I'd say this is an ops change primarily. You can add multiple categories in the parentheses separated by commas like this: `(enterprise change, ops change)`.
___
*[`pkg/ccl/oidcccl/settings.go` line 42 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/125767#-O0nkr9L1RksHOASMAlA:-O0nkr9L1RksHOASMAlB:b-uepmsn) ([raw file](https://github.com/cockroachdb/cockroach/blob/179b684adb845b11f7da8308a0c9cc1d94965f7a/pkg/ccl/oidcccl/settings.go#L42)):*
> ```Go
> 	OIDCGenerateClusterSSOTokenSQLHostSettingName  = baseOIDCSettingName + "generate_cluster_sso_token.sql_host"
> 	OIDCGenerateClusterSSOTokenSQLPortSettingName  = baseOIDCSettingName + "generate_cluster_sso_token.sql_port"
> 	oidcAuthClientTimeoutSettingName               = baseOIDCSettingName + "client_timeout"
> ```

nit: capitalization to match the rest
___
*[`pkg/ccl/oidcccl/authentication_oidc_test.go` line 574 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/125767#-O0nl2lFEy09D7ghuIaY:-O0nl2lFEy09D7ghuIaZ:b2kdigm) ([raw file](https://github.com/cockroachdb/cockroach/blob/179b684adb845b11f7da8308a0c9cc1d94965f7a/pkg/ccl/oidcccl/authentication_oidc_test.go#L574)):*
> ```Go
> 	require.NotNil(t, oidcMgr.oauth2Config)
> 	require.NotNil(t, oidcMgr.httpClient)
> 	require.Equal(t, clientTimeout, oidcMgr.httpClient.Timeout)
> ```

The test you've added does not actually test that the timeout is honored. In order to test this, your handler should never return and force the OIDC code to trigger the timeout.

Here's an example of what I'm talking about:

```
func TestOIDCProviderInitialization(t *testing.T) {
	defer leaktest.AfterTest(t)()
	defer log.Scope(t).Close(t)

	// Initiate a test server to handle `./well-known/openid-configuration`.
	testServer := httptest.NewUnstartedServer(nil)
	testServerURL := "http://" + testServer.Listener.Addr().String()

	waitChan := make(chan (struct{}))

	mux := http.NewServeMux()
	mux.HandleFunc(
		"GET /.well-known/openid-configuration",
		func(w http.ResponseWriter, r *http.Request) {
			<-waitChan
		},
	)

	testServer.Config = &http.Server{
		Handler: mux,
	}
	testServer.Start()
	defer testServer.Close()

	// Initialize the OIDC manager.
	clientTimeout := time.Millisecond
	oidcConf := oidcAuthenticationConf{
		providerURL:   testServerURL,
		clientTimeout: clientTimeout,
	}
	_, err := NewOIDCManager(context.Background(), oidcConf, "redirectURL", []string{})
	var urlError *url.Error
	require.ErrorAs(t, err, &urlError)
	require.True(t, urlError.Timeout())

	waitChan <- struct{}{}
	close(waitChan)
}
```

**Review by pritesh-lahoti — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125767)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @souravcrl)

___
*[`-- commits` line 2 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/125767#-O0nkQJG4vgSICLCxUfd):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

nit: just a stylistic nit that the title is a little vague. I'd use something like: `add configurable timeouts to HTTP clients`.
</blockquote></details>

Addressed
___
*[`-- commits` line 12 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/125767#-O0nkI0i9rNPc_1AyjmH):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

nit: I'd say this is an ops change primarily. You can add multiple categories in the parentheses separated by commas like this: `(enterprise change, ops change)`.
</blockquote></details>

Addressed
___
*[`pkg/ccl/oidcccl/settings.go` line 42 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/125767#-O0nkr9L1RksHOASMAlA:-O0nvzhd1kmSfDoD-epT:bql237d) ([raw file](https://github.com/cockroachdb/cockroach/blob/179b684adb845b11f7da8308a0c9cc1d94965f7a/pkg/ccl/oidcccl/settings.go#L42)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

nit: capitalization to match the rest
</blockquote></details>

These constants are not used outside of this package, hence need not be exported.
___
*[`pkg/ccl/oidcccl/authentication_oidc_test.go` line 574 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/125767#-O0nl2lFEy09D7ghuIaY:-O0ojQpM2kR3fn-6XolB:b-vjtmw0) ([raw file](https://github.com/cockroachdb/cockroach/blob/179b684adb845b11f7da8308a0c9cc1d94965f7a/pkg/ccl/oidcccl/authentication_oidc_test.go#L574)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

The test you've added does not actually test that the timeout is honored. In order to test this, your handler should never return and force the OIDC code to trigger the timeout.

Here's an example of what I'm talking about:

```
func TestOIDCProviderInitialization(t *testing.T) {
	defer leaktest.AfterTest(t)()
	defer log.Scope(t).Close(t)

	// Initiate a test server to handle `./well-known/openid-configuration`.
	testServer := httptest.NewUnstartedServer(nil)
	testServerURL := "http://" + testServer.Listener.Addr().String()

	waitChan := make(chan (struct{}))

	mux := http.NewServeMux()
	mux.HandleFunc(
		"GET /.well-known/openid-configuration",
		func(w http.ResponseWriter, r *http.Request) {
			<-waitChan
		},
	)

	testServer.Config = &http.Server{
		Handler: mux,
	}
	testServer.Start()
	defer testServer.Close()

	// Initialize the OIDC manager.
	clientTimeout := time.Millisecond
	oidcConf := oidcAuthenticationConf{
		providerURL:   testServerURL,
		clientTimeout: clientTimeout,
	}
	_, err := NewOIDCManager(context.Background(), oidcConf, "redirectURL", []string{})
	var urlError *url.Error
	require.ErrorAs(t, err, &urlError)
	require.True(t, urlError.Timeout())

	waitChan <- struct{}{}
	close(waitChan)
}
```
</blockquote></details>

Added a new test. Thank you!

**Review by dhartunian — APPROVED** [review]
:lgtm: thanks for making all the changes

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125767#-:-O0p_2LgCLBv0H_6cmWU:bidujcv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @pritesh-lahoti and @souravcrl)

___
*[`pkg/ccl/oidcccl/settings.go` line 42 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/125767#-O0nkr9L1RksHOASMAlA:-O0pZxQDBbdZNPuESefb:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/179b684adb845b11f7da8308a0c9cc1d94965f7a/pkg/ccl/oidcccl/settings.go#L42)):*
<details><summary><i>Previously, pritesh-lahoti wrote…</i></summary><blockquote>

These constants are not used outside of this package, hence need not be exported.
</blockquote></details>

:thumbsup:

**Review by pritesh-lahoti — COMMENTED** [review]
FYI - Had to rename the cluster setting from `server.oidc_authentication.client_timeout` to `server.oidc_authentication.client.timeout` to resolve lint validation errors.
Also, fixed the logictest as a result of introduction of this cluster setting.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125767#-:-O0s92FG5ama86ahLGI9:b4p8ecn)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @dhartunian and @souravcrl)

**Comment by pritesh-lahoti** [comment]
TYFTRs!
bors r=dhartunian,souravcrl
